lib/hotel.rb

```diff
@@ -0,0 +1,102 @@
+module Hotel
+  class Hotel
+    attr_reader :id, :rooms, :reservations, :hotel_blocks
+
+    def initialize(id:, rooms:, reservations: nil)
+      @id = id
+      @rooms = rooms
+      @reservations ||= reservations
+      @hotel_blocks = []
+    end
+
+    def reserve_room(start_date, end_date, room_id, block_id = nil, discount = nil)
+      #check if room available during those days
+      rooms_available = available_rooms(start_date, end_date)
+      raise ArgumentError if !(rooms_available.include?(room_id)) && block_id == nil
+      # find room object with room ID
+      room = find_room(room_id)
+      new_reservation = Reservation.new(id: create_reservation_id, start_date: start_date, end_date: end_date, room: room, block_id: block_id, discount: discount)
+      # p new_reservation.block_id
+      add_reservation(new_reservation)
+      room.add_reservation(new_reservation)
+
+      return new_reservation
+    end
+
+    def create_reservation_id
+      reservation_id = reservations.length + 1
+      return reservation_id
+    end
+
+    def create_block_id
+      block_id = hotel_blocks.length + 1
+      return block_id
+    end
+
+    def find_room(room_ID)
+      @rooms.detect { |room| room.id == room_ID }
+    end
+
+    def add_reservation(reservation)
+      reservations << reservation
+    end
+
+    def access_reservations(date)
+      list_reservations = @reservations.select { |reservation| (Range.new(reservation.start_date, reservation.end_date)).include?(date) }
+      return list_reservations.map { |reservation| reservation }
+    end
```

Comment:
On line 45 you do a lot of work with the internals of the reservation. I know we hadn't read POODR ch 4 yet when you wrote this code, but this is exactly what Metz is talking about when she says you should "ask for _what_ instead of telling _how_". A cleaner approach might be to add a `.includes_date?` method to reservation, then use it here:

```ruby
return @reservations.select { |res| res.include_date?(date) }
```

Also the `map` on line 46 is not needed.